core/src/main/java/org/apache/calcite/adapter/enumerable/RexToLixTranslator.java

```diff
@@ -727,11 +736,38 @@ private Expression translateParameter(RexDynamicParam expr,
     if (storageType == null) {
       storageType = typeFactory.getJavaClass(expr.getType());
     }
-    return nullAs.handle(
-        convert(
-            Expressions.call(root, BuiltInMethod.DATA_CONTEXT_GET.method,
-                Expressions.constant("?" + expr.getIndex())),
-            storageType));
+
+    Expression expression = Expressions.call(root, BuiltInMethod.DATA_CONTEXT_GET.method,
+            Expressions.constant("?" + expr.getIndex()));
+
+    Set<SqlTypeName> toIntegerTypes = Sets.newHashSet(
+            DATE,
+            TIME,
+            TIME_WITH_LOCAL_TIME_ZONE,
+            INTERVAL_YEAR,
+            INTERVAL_YEAR_MONTH,
+            INTERVAL_MONTH);
+
+    Expression midExpress = expression;
+    if (toIntegerTypes.contains(expr.getType().getSqlTypeName())) {
+      SqlTypeName t = expr.getType().getSqlTypeName();
+
+      switch (t) {
+      case DATE:
+        midExpress = Expressions.call(
+                BuiltInMethod.STRING_TO_DATE_ALLOW_NULLABLE.method,
+                convert(expression, String.class));
+        break;
+      case TIME:
+        midExpress = Expressions.call(
+                BuiltInMethod.STRING_TO_TIME_ALLOW_NULLABLE.method,
+                convert(expression, String.class));
+        break;
+      default:
+        //do nothing, if can't support now, just throw exception later
```

Comment:
Maybe we should throw an exception "cast to xxx type is not supported yet!"

Reply:
As I mentioned in "Conversation" tag, seems this PR just fixes the casting to type of DATE/TIME, what about other common types? e.g. cast(String as Integer) and so on ..... Yes I noticed this switch-default, but seems that in most cases exception will be thrown

Reply:
What about `TIMESTAMP`?